File: cpp/HandOfStraights.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <queue>

using namespace std;
class Solution {
public:
    bool isNStraightHand(vector<int>& hand, int groupSize) {
        int n = hand.size();
        // last group cant be created if the size of hand is not divisible by group size
        if(n % groupSize != 0){
            return false;
        }

        unordered_map<int, int> countMap;
        priority_queue<int, vector<int>, greater<int>> minHeap;
        for(int i = 0; i < n; i++){
            countMap[hand[i]] += 1;
        }
        for(auto h : countMap){
            minHeap.push(h.first);
        }
        while(!minHeap.empty()){
            int top = minHeap.top();

            for(int i = top; i < top + groupSize; i++){
                // if the key is not in hashMap we return false since we cant create a group
                if(!(countMap.count(i))){
                    return false;
                }
                countMap[i] -= 1;
                if(countMap[i] == 0){
                    // we can only pop the value if its the smallest value in the heap, which is always on top in the minHeap
                    if(i != minHeap.top()){
                        return false;
                    }
                    minHeap.pop();
                }
            }
        }
        return true;
    }
};

File: python/HandOfStraights.py
from typing import List
import heapq

class Solution:
    def isNStraightHand(self, hand: List[int], groupSize: int) -> bool:
        # if the size of array cant be divided with remainder 0, then return False, since we cant create groups
        if len(hand) % groupSize != 0:
            return False
        
        # counter hashMap for all values in hand
        countMap = {}

        for i in hand:
            if i in countMap:
                countMap[i] += 1
            else:
                countMap[i] = 1

        # minHeap to check the least freq element
        minHeap = list(countMap.keys())
        heapq.heapify(minHeap)

        while minHeap:
            top = minHeap[0]

            # check all the numbers stating from top of heap up to the last consequative element that needsto be in the group
            for i in range(top, top + groupSize):
                # if the elem is not in hashMap, that means we cant make a group
                if i not in countMap:
                    return False
                countMap[i] -= 1
                # we can only pop the value from heap if its the least frequent
                if countMap[i] == 0:
                    if i != minHeap[0]:
                        return False
                    heapq.heappop(minHeap)
        
        return True